src/Features/Core/Portable/IntroduceUsingStatement/AbstractIntroduceUsingStatementCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,291 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.CodeAnalysis.Utilities;
+using static Microsoft.CodeAnalysis.CodeActions.CodeAction;
+
+namespace Microsoft.CodeAnalysis.IntroduceUsingStatement
+{
+    internal abstract class AbstractIntroduceUsingStatementCodeRefactoringProvider<TStatementSyntax, TLocalDeclarationSyntax, TBlockSyntax> : CodeRefactoringProvider
+        where TStatementSyntax : SyntaxNode
+        where TLocalDeclarationSyntax : TStatementSyntax
+        where TBlockSyntax : SyntaxNode
+    {
+        protected abstract SyntaxList<TStatementSyntax> GetStatements(TBlockSyntax blockSyntax);
+
+        protected abstract TBlockSyntax WithStatements(TBlockSyntax blockSyntax, SyntaxList<TStatementSyntax> statements);
+
+        protected abstract TStatementSyntax CreateUsingStatement(TLocalDeclarationSyntax declarationStatement, SyntaxTriviaList sameLineTrivia, SyntaxList<TStatementSyntax> statementsToSurround);
+
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var document = context.Document;
+            var span = context.Span;
+
+            var (declarationSyntax, _) =
+                await FindDisposableLocalDeclaration(document, span, context.CancellationToken).ConfigureAwait(false);
+
+            if (declarationSyntax != null)
+            {
+                context.RegisterRefactoring(new MyCodeAction(
+                    FeaturesResources.Introduce_using_statement,
+                    cancellationToken => IntroduceUsingStatementAsync(document, span, cancellationToken)));
+            }
+        }
+
+        private static async Task<(TLocalDeclarationSyntax, ILocalSymbol)> FindDisposableLocalDeclaration(Document document, TextSpan selection, CancellationToken cancellationToken)
+        {
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+            var declarationSyntax =
+                root.FindNode(selection)?.GetAncestor<TLocalDeclarationSyntax>()
+                ?? root.FindTokenOnLeftOfPosition(selection.End).GetAncestor<TLocalDeclarationSyntax>();
```

Comment:
do you need both?  is either one not good enough?

Reply:
A bunch of tests fail. Particularly interesting are these three:

https://github.com/dotnet/roslyn/blob/6b29eb5d88e2b42e449e812d9f273d27f9bd53f2/src/EditorFeatures/CSharpTest/IntroduceUsingStatement/IntroduceUsingStatementTests.cs#L52-L120

Simpler solutions pick the wrong statement to refactor in for the last two tests or don't work at all for the first test.

Reply:
@CyrusNajmabadi Sorry, I replied without seeing your comment somehow. 

> Why does `root.FindTokenOnLeftOfPosition(selection.Start).GetAncestor` not work for all three of those cases?

It would pick the wrong statement to refactor here:

```c#
            await TestInRegularAndScriptAsync(
@"class C
{
    void M(System.IDisposable disposable)
    {
        var ignore = disposable;
        [||]var name = disposable;
    }
}",
@"class C
{
    void M(System.IDisposable disposable)
    {
        var ignore = disposable;
        using (var name = disposable)
        {
        }
    }
}");
```

Reply:
It should stay on the correct line. FindNode works great. Any ideas for simplifying?

Reply:
if this works well, then stick with it :)